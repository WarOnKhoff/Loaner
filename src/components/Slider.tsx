import React, { FC, ReactElement, ChangeEvent } from "react";

interface IProps {
  minValue: number;
  maxValue: number;
  currentValue: number;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  selector?: string;
}
const Slider: FC<IProps> = ({
  minValue,
  maxValue,
  onChange,
  currentValue,
  selector
}: IProps): ReactElement => {
  const getSliderBackground = (): string =>
    `linear-gradient(90deg, #DA2C38 ${(currentValue / maxValue) *
      100}%, #747572 ${(currentValue / maxValue) * 100}%)`;
  return (
    <input
      type="range"
      min={minValue}
      max={maxValue}
      value={currentValue}
      style={{
        background: getSliderBackground()
      }}
      onChange={onChange}
      className={`slider ${selector}`}
      data-testid={selector}
    />
  );
};

export default React.memo(Slider);
