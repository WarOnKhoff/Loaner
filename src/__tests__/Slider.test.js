import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Slider from "../components/Slider";

const setup = props => {
  const { getByTestId, container } = render(<Slider {...props} />);
  return {
    container,
    sliderInput: getByTestId(props.selector)
  };
};

const mocks = {
  minValue: 1,
  maxValue: 100,
  onChange: jest.fn(),
  currentValue: 5,
  selector: "test-selector"
};

describe("Components: Slider", () => {
  test("Renders correctly", () => {
    const { container } = setup(mocks);
    expect(container).toMatchSnapshot();
  });
  test("Should be type of range input", () => {
    const { sliderInput, container } = setup(mocks);
    expect(sliderInput.getAttribute("type")).toBe("range");
    // expect(container).toBe("range");
  });
  test("Should receive minValue from props", () => {
    const { sliderInput } = setup(mocks);
    expect(sliderInput).toHaveAttribute("min");
    // mocks.minValue.toString() is used because of default input specificity
    // It keeps and provides string value
    expect(sliderInput.getAttribute("min")).toBe(mocks.minValue.toString());
  });
  test("Should receive maxValue from props", () => {
    const { sliderInput } = setup(mocks);
    expect(sliderInput).toHaveAttribute("max");
    expect(sliderInput.getAttribute("max")).toBe(mocks.maxValue.toString());
  });
  test("Should receive currentValue from props", () => {
    const { sliderInput } = setup(mocks);
    expect(sliderInput).toHaveAttribute("value");
    expect(sliderInput).toHaveValue(mocks.currentValue.toString());
  });
  test("Should keep proper className with selector from props", () => {
    const { sliderInput } = setup(mocks);
    expect(sliderInput.className).toBe(`slider ${mocks.selector}`);
  });
  test("Should fire onChange callback when is being changed", () => {
    const { sliderInput } = setup(mocks);
    sliderInput.onchange = mocks.onChange;
    fireEvent.change(sliderInput);
    expect(mocks.onChange).toHaveBeenCalled();
  });
});
