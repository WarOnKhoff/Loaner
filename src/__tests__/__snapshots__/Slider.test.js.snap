// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Slider Renders correctly 1`] = `
<div>
  <input
    class="slider test-selector"
    data-testid="test-selector"
    max="100"
    min="1"
    type="range"
    value="5"
  />
</div>
`;
